src/cls-ah.js

```diff
@@ -0,0 +1,118 @@
+/**
+ * Copyright 2017 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+const asyncHook = require('async_hooks');
+
+const wrappedSymbol = Symbol('context_wrapped');
+const contexts = {};
+var current = {};
+
+const hook = asyncHook.createHook({init, before, after, destroy});
+hook.enable();
+
+function Namespace() {}
+
+Namespace.prototype.get = function(k) {
+  return current ? current[k] : null;
+};
+
+Namespace.prototype.set = function(k, v) {
+  if (current) {
+    current[k] = v;
+  }
+};
+
+Namespace.prototype.run = function(fn) {
+  const oldContext = current;
+  current = {};
+  const res = fn();
+  current = oldContext;
+  return res;
+};
+
+Namespace.prototype.runAndReturn = function(fn) {
+  return this.run(fn);
+};
+
+Namespace.prototype.bind = function(cb) {
+  if (cb[wrappedSymbol] || !current) {
+    return cb;
+  }
+  const boundContext = current;
+  const wrapped = cb.length === 0 ? function() {
+    const oldContext = current;
+    current = boundContext;
+    const res = cb.apply(this, arguments);
+    current = oldContext;
+    return res;
+  } : function(a) {
```

Comment:
The new approach uses `Object#defineProperty`. @bmeurer are there any expected performance implications of redefining a function's length? I played around with some microbenchmarks that didn't show any impact but am concerned the microbenchmarks may not show everything.